neps/nep-0541.md

```diff
@@ -0,0 +1,196 @@
+    # store the actual receipt with index in the heap
+    store_receipt(DelayedReceipt(receipt, priority_index))
+```
+
+### Validator Reward
+
+Even though conceptually, a chunk producer is rewarded a portion of the priority fee for each transaction included in the chunk. We cannot implement this naively due to two reasons:
+- the chunk producer account may be on a different shard;
+- the staking logic dictates that validator reward be applied to validator accounts at the beginning of an epoch.
+
+To make this computation easy, we add a new field `total_priority_fee: u64` to chunk header that stores the sum of all priority fees of transactions included in this chunk. Then, at the end of an epoch, we tally the cumulative priority fees for each chunk producer (in implementation it can be updated on each block) and then add it to their staking reward.
+If a chunk producer does not meet the online requirement for an epoch, they are not eligible for reward based on priority fees for the entire epoch.
+From the perspective of total supply of the protocol, the priority fee is first all burnt when a transaction is processed and then at the beginning of the next epoch, some tokens are minted to reward chunk producers for including those transactions.
+
+One thing to note here is that since `total_priority_fee` should be within `u64` range, we should cap the maximum of `priority_fee` for each transaction at ~1800N (this generously assumes that a chunk may contain up to 10,000 transactions).
+
+
+## Security Implications
+
+[Explicitly outline any security concerns in relation to the NEP, and potential ways to resolve or mitigate them. At the very least, well-known relevant threats must be covered, e.g. person-in-the-middle, double-spend, XSS, CSRF, etc.]
+
+## Alternatives
+
+One clear alternative is to not have FIFO processing at all and make all receipt processing purely based on priority. This, however, would create the problem that some receipts without priority can be indefinitely delayed when receipts with higher priority keeps coming in.
+Then there is the question of whether those receipts should pay for its storage in the state since there is no bound on when they may get processed.
+Furthermore, since there is no way to cancel a receipt after it is created, even if a user notices that their receipt is stuck in the queue for a long time and would like to cancel,
+they won't be able to do it.
+
+## Future possibilities
+
+This NEP should be combined with [NEP-513](https://github.com/near/NEPs/pull/539). They together redefine how congestion is handled and how users can still send transactions during congestion. It is possible to explore more complex mechanism on priority fees when there is congestion. For example, the protocol could require that transactions to a congested shard must attach a priority fee and even place a minimum on the priority fee based on previous chunk's priority fees.
```

Comment:
Oh and how are we dealing with priority in the outgoing buffers that NEP-539 currently proposes?

I would probably suggest that, at least initially, draining the outgoing buffers should follow a strict FIFO order. The priority already helped to get inside faster, so maybe the "cross-shard delay" can be fair.

Alternatively, we can add more priority queues (one extra for each receiving shard, per shard) and give truly fast speed even to cross-contract calls during congestion. It would certainly be a better user experience, since otherwise one can still wait a long time for a cross-contract call during congestion, no matter how much one pays.

I just feel perhaps an iterative approach would be better than changing so much all at once, without the experience of how congestion control works in practice.

Reply:
yeah I think that makes sense